File: test/tests/nodes_test.dart
// @Timeout(Duration(seconds: 10))
import 'dart:async';

import 'package:dartz/dartz.dart';
import 'package:sangre/sangre.dart';
import 'package:test/test.dart';

void main() {
  test('Growling list source', () async {
    final gls = await GrowingListSource(3);
    expect(
      gls.stream,
      emitsInOrder([
        [0],
        [0, 1],
        [0, 1, 2],
        emitsDone
      ]),
    );
  });

  test('Combine growing list source with count operator', () async {
    final chain = await NodeOperator1Input(
      (Iterable a) async => a.length,
      await GrowingListSource(3),
    );
    expect(
      chain.stream,
      emitsInOrder([1, 2, 3, emitsDone]),
    );
  }, timeout: Timeout(Duration(seconds: 3)));

  test('Combine growing list source with an async operator', () async {
    final chain = await NodeOperator1Input(
      (Iterable a) => Future.delayed(
        Duration(seconds: 1),
        () => a.length,
      ),
      await GrowingListSource(3),
    );
    expect(
      chain.stream,
      emitsInOrder([1, 2, 3, emitsDone]),
    );
  });

  test('Combine two growing list source with count operator', () async {
    final chain = await NodeOperator2Input(
      (Iterable a, Iterable b) async => a.length + b.length,
      await GrowingListSource(3),
      await GrowingListSource(4),
    );
    expect(
      chain.stream,
      emitsInOrder([2, 3, 4, 5, 6, 7, emitsDone]),
    );
  });

  test('React properly to source change', () async {
    // This test has no semantic meaning, but tests source propagation
    final ListSource<Tuple2<int, String>> usersDBSource =
        await ListSource<Tuple2<int, String>>();

    for (var i = 0; i < 4; i++)
      usersDBSource.insertRow(
        Tuple2(i, randomString()),
      );

    final intSource = await GrowingListSource(3);
    final chain = await NodeOperator2Input(
      (List<Tuple2<int, String>> users, Iterable b) async =>
          users.length + b.length,
      usersDBSource,
      intSource,
    );

    expect(
      chain.stream,
      emitsInOrder([5, 6, 7]),
    );

    await intSource.streamController.last;

    usersDBSource.insertRow(
      Tuple2(4, randomString()),
    );
    expect(
      chain.stream,
      emitsInOrder([7, 8]),
    );
  });

  test('Node chaining without await', () async {
    final chain = await NodeOperator1Input(
      (Iterable a) async => a.length,
      GrowingListSource(3),
    );
    expect(
      chain.stream,
      emitsInOrder([1, 2, 3, emitsDone]),
    );
  });
}

File: test/nodes.test.ts

import _ from 'lodash'
import { describe, it } from 'mocha'
import { delayed, GrowingListSource, NodeOperator1Input, NodeOperator2Input } from '../src/index'
import { expectNodeToEmitInOrder } from './index.test'

describe("Sources", async function () {
    it('Growling list source', async function () {
        const source = await new GrowingListSource(3)
        await expectNodeToEmitInOrder(source, [
            [0],
            [0, 1],
            [0, 1, 2],
        ])
    })
})


describe("Operators", async function () {
    it('Combine growing list source with count operator', async function () {
        const chain = await new NodeOperator1Input(
            async (values) => values.length,
            new GrowingListSource(3),
        )
        await expectNodeToEmitInOrder(chain, [1, 2, 3])
    })

    it('Combine growing list source with count operator (with await)', async function () {
        const chain = await new NodeOperator1Input(
            async (values) => values.length,
            await new GrowingListSource(3),
        )
        await expectNodeToEmitInOrder(chain, [1, 2, 3])
    })

    it('Combine fast growing list source with an async operator', async function () {
        const chain = await new NodeOperator1Input(
            async (a) => await delayed(50, () => a.length * 2),
            new GrowingListSource(3),
        )
        await expectNodeToEmitInOrder(chain, [2, 4, 6])
    })

    it('Combine slow growing list source with an async operator', async function () {
        const chain = await new NodeOperator1Input(
            async (a) => await delayed(1500, () => a.length * 2),
            new GrowingListSource(3),
        )
        await expectNodeToEmitInOrder(chain, [2, 4, 6])
    })

    it('Combine two growing list source with count operator', async function () {
        const chain = await new NodeOperator2Input(
            async (a, b) => a.length + b.length,
            new GrowingListSource(3),
            new GrowingListSource(4),
        )
        await expectNodeToEmitInOrder(
            chain,
            [2, 3, 4, 5, 6, 7]
        )
    })

    it('React properly to source change', async function () {
        const source = await new GrowingListSource(3)
        const chain = await new NodeOperator1Input(
            async (a) => _.sum(a),
            source,
        )

        delayed(source.intervalInMs * 5, () => source.insertRow.bind(source)(5))
        await expectNodeToEmitInOrder(chain, [0, 1, 3, 8])
    })
})
